Firestore/Swift/Tests/Codable/CodableDocumentTests.swift

```diff
@@ -0,0 +1,403 @@
+    XCTAssertNotNil(model2)
+  }
+
+  // Inspired by https://github.com/firebase/firebase-android-sdk/blob/master/firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java
+  func testBeans() {
+    struct Model: Codable {
+      let s: String
+      let d: Double
+      let f: Float
+      let l: CLongLong
+      let i: Int
+      let b: Bool
+      let sh: CShort
+      let byte: CChar
+      let uchar: CUnsignedChar
+      let ai: [Int]
+      let si: [String]
+      let caseSensitive: String
+      let casESensitive: String
+      let casESensitivE: String
+    }
+    let model = Model(
+      s: "abc",
+      d: 123,
+      f: -4.321,
+      l: 1_234_567_890_123,
+      i: -4444,
+      b: false,
+      sh: 123,
+      byte: 45,
+      uchar: 44,
+      ai: [1, 2, 3, 4],
+      si: ["abc", "def"],
+      caseSensitive: "aaa",
+      casESensitive: "bbb",
+      casESensitivE: "ccc"
+    )
+    let dict = [
+      "s": "abc",
+      "d": 123,
+      "f": -4.321,
+      "l": 1_234_567_890_123,
+      "i": -4444,
+      "b": false,
+      "sh": 123,
+      "byte": 45,
+      "uchar": 44,
+      "ai": [1, 2, 3, 4],
+      "si": ["abc", "def"],
+      "caseSensitive": "aaa",
+      "casESensitive": "bbb",
+      "casESensitivE": "ccc",
+    ] as [String: Any]
+
+    let model2 = try! Firestore.Decoder().decode(Model.self, from: dict)
+    XCTAssertEqual(model.s, model2.s)
+    XCTAssertEqual(model.d, model2.d)
+    XCTAssertEqual(model.f, model2.f)
+    XCTAssertEqual(model.l, model2.l)
+    XCTAssertEqual(model.i, model2.i)
+    XCTAssertEqual(model.b, model2.b)
+    XCTAssertEqual(model.sh, model2.sh)
+    XCTAssertEqual(model.byte, model2.byte)
+    XCTAssertEqual(model.uchar, model2.uchar)
+    XCTAssertEqual(model.ai, model2.ai)
+    XCTAssertEqual(model.si, model2.si)
+    XCTAssertEqual(model.caseSensitive, model2.caseSensitive)
+    XCTAssertEqual(model.casESensitive, model2.casESensitive)
+    XCTAssertEqual(model.casESensitivE, model2.casESensitivE)
+
+    let encodedDict = try! Firestore.Encoder().encode(model)
+    XCTAssertEqual(encodedDict["s"] as! String, "abc")
+    XCTAssertEqual(encodedDict["d"] as! Double, 123)
+    XCTAssertEqual(encodedDict["f"] as! Float, -4.321)
+    XCTAssertEqual(encodedDict["l"] as! CLongLong, 1_234_567_890_123)
+    XCTAssertEqual(encodedDict["i"] as! Int, -4444)
+    XCTAssertEqual(encodedDict["b"] as! Bool, false)
+    XCTAssertEqual(encodedDict["sh"] as! CShort, 123)
+    XCTAssertEqual(encodedDict["byte"] as! CChar, 45)
+    XCTAssertEqual(encodedDict["uchar"] as! CUnsignedChar, 44)
+    XCTAssertEqual(encodedDict["ai"] as! [Int], [1, 2, 3, 4])
+    XCTAssertEqual(encodedDict["si"] as! [String], ["abc", "def"])
+    XCTAssertEqual(encodedDict["caseSensitive"] as! String, "aaa")
+    XCTAssertEqual(encodedDict["casESensitive"] as! String, "bbb")
+    XCTAssertEqual(encodedDict["casESensitivE"] as! String, "ccc")
+  }
+
+  func testCodingKeys() {
```

Comment:
This intentionally won't `roundTrip`, right? Maybe it's worth adding a test to verify it throws

Reply:
Why should it throw?

Reply:
Ahh sorry - I think I read this before seeing that `init(from decoder: Decoder)` uses default values for the properties that don't have associated coding keys.